pydarn/io/superdarn.py

```diff
@@ -0,0 +1,706 @@
+
+    # helper function that could be used parallelization
+    def _read_darn_record(self, format_fields: List[dict]):
+        """
+        Read SuperDARN DMAP records from the DMAP byte array. Several SuperDARN
+        field checks are done to insure the integrity of the file.
+
+        Parameters
+        ----------
+        format_fields : List[dict]
+            Is a list of dictionaries for the fields that are possible in a
+            SuperDARN file type
+
+        Raises:
+        -------
+        SuperDARNFieldMissingError - when a field is missing from the SuperDARN
+                                file/stream type
+        SuperDARNExtraFieldError - when an extra field is present in the
+                                SuperDARN file/stream type
+        SuperDARNDataFormatTypeError - when a field has the incorrect
+                                field type for the SuperDARN file/stream type
+
+        See Also
+        --------
+        read_record() - inherited from DmapRead
+        """
+        record = self.read_record()
+        DarnUtilities.missing_field_check(format_fields, record, self.rec_num)
+        DarnUtilities.extra_field_check(format_fields, record, self.rec_num)
+        DarnUtilities.incorrect_types_check(format_fields, record, self.rec_num)
+        self._dmap_records.append(record)
+
+    def _read_darn_records(self, format_fields: List[dict]):
+        """
+        loops over the bytes in the in the SuperDARN byte array and
+        calls the helper method read the SuperDARN records from the file/stream
+
+        Parameters
+        ----------
+        format_fields : List[dict]
+            Is a list of dictionaries for the fields that are possible in a
+            SuperDARN file type
+
+        See Also
+        --------
+        _read_darn_record
+        """
+        self.rec_num = 0  # record number, for exception info
+        while self.cursor < self.dmap_end_bytes:
+            self._read_darn_record(format_fields)
+            self.rec_num += 1
+
+    def read_iqdat(self) -> List[dict]:
+        """
+        Reads iqdat DMAP file/stream
+
+        Returns
+        -------
+        dmap_records : List[dict]
+            DMAP record of the Iqdat data
+        """
+        pydarn_log.debug("Reading Iqdat file: {}".format(self.dmap_file))
+        file_struct_list = [superdarn_formats.Iqdat.types]
+        self._read_darn_records(file_struct_list)
+        return self._dmap_records
+
+    def read_rawacf(self) -> List[dict]:
+        """
+        Reads Rawacf DMAP file/stream
+
+        Returns
+        -------
+        dmap_records : List[dict]
+            DMAP record of the Rawacf data
+        """
+        pydarn_log.debug("Reading Rawacf file: {}".format(self.dmap_file))
+
+        file_struct_list = [superdarn_formats.Rawacf.types]
+        self._read_darn_records(file_struct_list)
+        return self._dmap_records
+
+    def read_fitacf(self) -> List[dict]:
```

Comment:
Most of the documentation on this is actually in the write functions because that is where field checking was done first. As you saw (maybe not fully understand) in the missing field check I pass in multiple subsets (granted they will probably evolve over time as I haven't checked for everything under the uns since RST doesn't state which fields are optional) that then get checked on the recordset to find any missing fields but not missing subsets.

Reply:
Is it possible to maybe copy that documentation here, or maybe refer to where one should look. I think everything else was pretty good. This particular piece of code required lots of looking around though to understand.

Reply:
I think we are getting a bit pedantic about this comment/documentation now. The new commits show newly updated comments around this function and others pertaining to this edge case. 

Plus I gave an explanation in the above comment so I think this is satisfactory now :) 
Worthwileness vs. conflict :)